### ISSUE TITLE
Loading ROM's into the code segment size

### ISSUE THREAD

**Issue by Sworddragon** [mamedev/mame#2380]
I have thought about this topic for a while and I'm assuming that MAME interprets arbitrary data (like BIOS ROM's and game ROM's) like probably the most emulators do. This interpretation by emulating for example the cpu instructions has commonly a very high overhead and is the main cause why emulation is not that fast.

Since this sort of emulation implies that the emulator is capable of doing the necessary decompiling tasks by reading and interpreting the ROM and the compiling tasks by executing the correct host instructions wouldn't it be much more effective to do a native recompilation (including the virtual machine capabilities) and to load the result in the code segment size either of the MAME process or into an own process? Of course this introduces potential massive security concerns but these can be resolved by also compiling all the necessary sandbox capabilities into the resulting code (basically what #2378 for MAME's current state requests). Dependent how high the VM/sandbox overhead is and how well the compilation optimization is done this should be able to achieve a performance very close to native system performance.

Or is there a very specific reason to not to do such a sort of emulation?

**Comment by mizapf** [comment]
MAME has always been more than an attempt to make a visibly good emulation. An important part of the mission is to preserve the architectural concepts of the real systems, in particular the behaviour of the components. Ideally, by reading the implementations of the driver and the devices, you should be able to learn how e.g. a video processor draws its lines, how a disk controller chip goes through its state machine, or how wait states are generated.

MAME's performance could be significantly raised with a better concept for multithreading; currently, most work is done in a single thread.

**Comment by rb6502** [comment]
Back to the OP's major point: MAME does include what we call a "dynamic recompiler" (what most people call a "JIT") for specific high-speed CPUs: currently MIPS, SH-1, and SH-2.  This system batch-converts a subroutine at a time of the original code into native x64 as the program runs.

**Comment by DavidHaywood** [comment]
Indeed, the OP should also take into account that MAME also needs to reproduce the original timing of the machines very closely, and for many early machines that does actually mean a lot of the overhead is managing all that anyway - simply recompiling code into native blocks doesn't actually save you much if you can still only emulate one instruction at a time because you have to be absolutely sure everything runs in sync.

That's why MAME might currently have recompilers for things like MIPS, but doesn't have one for something like the Z80, which really needs to be moving in the opposite direction with sub-instruction level timing, breaking even the interpretor up into tiny interruptible segments etc.

**Comment by galibert** [comment]
Recompilation achieves a better performance than pure interpretation, but
nowhere near native performance.  Some of the elephants in the room:
- every memory access, or at least write, must be done, and memory access
decoding is a large cost in running an emulator
- subroutine calls write the return address to the stack usually, or more
often than not save registers to the stack
- unsolved jumps are expensive
- returning from a subroutine is an unsolved jump
- self-modifying code requires detection and recompilation buffer flushing

And that's only some of the problems...

  OG.

**Comment by Sworddragon** [comment]
> Back to the OP's major point: MAME does include what we call a "dynamic recompiler" (what most people call a "JIT") for specific high-speed CPUs: currently MIPS, SH-1, and SH-2. This system batch-converts each subroutine of the original code into native x86 or x64 as the program runs.

So the JIT compiler does create native machine code which is also directly executed by the CPU? Or is it just translated into a form of bytecode which needs still to be interpreted at the end?

**Comment by rb6502** [comment]
Yes.

As has become the standard for such things the initial CPU code is turned into a bytecode that's like an idealized x64, including some advanced bit manipulation instructions that e.g. PowerPC or Alpha had but Intel's never bothered with.

If you run with the compiler enabled on a CPU we don't have a backend for (ARM) then the bytecode is interpreted and the result is usually slower than just running a traditional interpreter,  But it does work, and it's great for debugging.

If you run with the compiler enabled on x86 or x64, then the bytecode is translated to target machine code and runs directly.  Note that no attempt is made to optimize the generated code, but just doing that is good for probably a 10 times speedup over a standard interpreter.  This is why interpreted CPUs in MAME bog down much above 25 MHz on top-end PCs while JITted ones can run 200-300 MHz.

**Comment by Sworddragon** [comment]
> If you run with the compiler enabled on a CPU we don't have a backend for (ARM) then the bytecode is interpreted and the result is usually slower than just running a traditional interpreter, But it does work, and it's great for debugging.

Shouldn't the bytecode still be faster than raw interpretation in this case?

So the CPU instructions from the ROM are being interpreted and turned into bytecode which is then either interpreted if MAME doesn't support it for the host CPU otherwise the bytecode is loaded into the code segment size and directly being executed. At least this is now how I think it works if I have not misunderstood something.


> Indeed, the OP should also take into account that MAME also needs to reproduce the original timing of the machines very closely, and for many early machines that does actually mean a lot of the overhead is managing all that anyway - simply recompiling code into native blocks doesn't actually save you much if you can still only emulate one instruction at a time because you have to be absolutely sure everything runs in sync.
>
> That's why MAME might currently have recompilers for things like MIPS, but doesn't have one for something like the Z80, which really needs to be moving in the opposite direction with sub-instruction level timing, breaking even the interpretor up into tiny interruptible segments etc.

Theoretically *pu limiter can and should normally be implemented no matter if the code is being interpreted or directly executed to get the most accuracy. Even if the host CPU is fast enough that it would make no difference in speed if the ROM is being interpreted or as machine code executed the direct executing would still be more efficient in terms of power consumption, host multitasking, etc. (also another host CPU might not be fast enough so that it would benefit from native execution).

**Comment by rb6502** [comment]
Interpreting the bytecode should in theory be faster than raw interpretation, but MIPS and SH are pretty trivial to decode (and were designed with that as a goal) so the instruction decode is generally about as simple as it is for the bytecode, and execution time is dominated by the actual execution.  Throw in the inefficiencies of the JIT framework vs. a purpose-built interpreter and you've got actual slowdown.

Obviously we limit the CPU speed even in JIT mode, but normally in JIT mode it's done at the end of a so-called "recompilation block", which is typically a subroutine.  Testing after each instruction eats a significant amount of the speedup you'd otherwise get from a JIT.

**Comment by DavidHaywood** [comment]
Right, my point was also more for if we start trying to optimize the generated code and lose all sense of what the original instructions were, you definitely can't expect multiple CPUs to run instructions in step if you've done that, although it is what some current emulators have to do I believe.

Code blocks also cause problems with external hardware accesses, if the timing is accurately emulated on another device, and expects a delay between accesses to that device but both accesses fall within a single recompiled block + optimized block it starts to get messy and the other emulated device can end up failing because it isn't happy with the timing of the accesses.  Anyway, that's drifting off topic.

**Comment by Sworddragon** [comment]
I prefer accuracy (and also security) over performance too and am a bit sceptic about the potential inaccuracy that current JIT codeblocks can cause but I get the point why they can make sense. Probably we need just different modes/options like interpreter-accuracy, interpreter-speed, JIT-accuracy, JIT-speed but that's something for another topic.

**Comment by galibert** [comment]
One thing that saves JITs/DRCs is that the emulated cpus that benefit from
them (mips, arm, sh, 386+) include caches and pipelines, which makes
instruction duration very contextual and hard to predict in the first
place.  As a consequence software usually didn't rely on precise
cycle-level timings.

  OG.

**Comment by Sworddragon** [comment]
That's a good point but there still might be the chance of race conditions for some software. Also in another scenario I'm thinking of the accuracy needed from an outer view of the emulated system. For example in competition-related speedrunning of video games where you have splits with an accuracy of 1/1000 of a second there might be a low chance that a race from an inaccurate JIT-compiler appears which might prevent a lag very close to the next millisecond resulting in having a recorded time of 1 millisecond faster than a physical system with the exact same execution would have. That could make it difficult to accept/allow in such scenarios mixed environments (emulator and physical system).

**Comment by angelosa** [comment]
Stale discussion, closing. How to make MAME fully multithreaded is probably described in libuv library #404
